config/configuration.go

```diff
@@ -0,0 +1,113 @@
+	"io/ioutil"
+	"net/url"
+)
+
+// we can have values set to default, except for
+//    scheduler { storage_uri }
+//    controller { compute_ca, compute_cert, identity_user, identity_password }
+//    launcher { compute_net, mgmt_net }
+//    image_service { url }
+//    identity_service { url }
+//
+// so we need to have at least those values set in our config
+//
+// TODO: proper validation of values set in yaml setup
+func validMinConf(conf *payloads.Configure) bool {
+	return (conf.Configure.Scheduler.ConfigStorageURI != "" &&
+		conf.Configure.Controller.ComputeCACert != "" &&
+		conf.Configure.Controller.ComputeCert != "" &&
+		conf.Configure.Controller.IdentityUser != "" &&
+		conf.Configure.Controller.IdentityPassword != "" &&
+		conf.Configure.Launcher.ComputeNetwork != "" &&
+		conf.Configure.Launcher.ManagementNetwork != "" &&
+		conf.Configure.ImageService.URL != "" &&
+		conf.Configure.IdentityService.URL != "")
+}
+
+func getConfiguration(conf_yaml []byte) (conf *payloads.Configure, err error) {
+	err = yaml.Unmarshal(conf_yaml, &conf)
+	if err != nil {
+		return nil, err
+	}
+	return
+}
+
+func getPayload(conf *payloads.Configure) (payload []byte, err error) {
+	if !validMinConf(conf) {
+		return nil, errors.New(
+			"minimal configuration is not met or yaml is malformed")
+	}
+	payload, err = yaml.Marshal(&conf)
+	if err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+func fillDefaults(conf *payloads.Configure) {
+	if conf.Configure.Scheduler.ConfigStorageType == "" {
+		conf.Configure.Scheduler.ConfigStorageType = "file"
+	}
+	if conf.Configure.Controller.ComputePort == 0 {
+		conf.Configure.Controller.ComputePort = 8774
+	}
+	if conf.Configure.ImageService.Type == "" {
+		conf.Configure.ImageService.Type = "glance"
+	}
+	if conf.Configure.IdentityService.Type == "" {
+		conf.Configure.IdentityService.Type = "keystone"
+	}
+}
+
+// handle the yaml configuration file read
+// only 'file' URI Scheme is supported at the moment
+// TODO: rework this more to check what's missing to
+// support etcd backend as well
+func loadFile(uri_str string) (yaml_conf []byte, err error) {
+	uri, err := url.Parse(uri_str)
+	if err != nil {
+		return nil, err
+	}
+	// check Scheme is supported
+	if uri.Scheme != "file" {
+		return nil, errors.New(
+			"configuration URI Scheme '" + uri.Scheme + "' not supported")
+	}
+	if uri.Path == "" {
+		return nil, errors.New("configuration URI path is empty")
+	}
+	yaml_conf, err = ioutil.ReadFile(uri.Path)
+	if err != nil {
+		return nil, err
+	}
+	return yaml_conf, nil
+}
+
+// GetConfiguration returns the payload to be set
+// in the SSNTP server. It needs an URI string
+// where it can read the configuration in yaml format.
+// If the URI is not valid or the yaml is either incomplete or
+// malformed, the payload recieved will be nil and err should
+// specify what the failure is
+func GetConfiguration(uri string) (payload []byte, err error) {
+	conf_yaml, err := loadFile(uri)
```

Comment:
So this needs a better abstraction. The theory of operation here is the following:
1. Define a configuration back-end Go interface
2. Implement a file back-end driver that implements the interface defined in 1. 
3. Parse the URI in order to gather the following pieces of information:
** The configuration backend URI
** The configuration back-end driver (file, etcd, ...)
4. Depending on the driver we inferred from the URI we will instantiate the specific back-end driver
5. Call into the back-end instance interface to fetch the actual configuration payload

In practice:

1. In config/driver.go:
```
type driver interface {
      fetchConfiguration(uri string) (payloads.Configure, error)
      storeConfiguration(payloads.Configure) (error)
}
```
2 In configuration/file.go
```
type file struct {
        file string
}

func (f *file) fetchConfiguration(uri string) (payloads.Configure, error) {
        [...]
}

func (f *file) storeConfiguration(payloads.Configure) (error) {
        [...]
}
```

4. in configuration/configuration.go:
```
func GetConfiguration(uri string) (payload []byte, err error) {
    [...]
    var d driver 
    if scheme == "file" {
          d = &file{}
    } else if scheme == "etcd" {
          d =  &etcd{}
    } else {
          error
    }

    configure, err := d.fetchConfiguration(uri)
    [...]
}
```